physicsnemo/distributed/shard_utils/unary_ops.py

```diff
@@ -0,0 +1,147 @@
+# SPDX-FileCopyrightText: Copyright (c) 2023 - 2024 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-FileCopyrightText: All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Unary operation helpers and dispatch wrappers for `ShardTensor`.
+
+This module provides:
+- A dispatch-level wrapper for `torch.unsqueeze` that preserves and adjusts
+  sharding metadata for `ShardTensor`.
+- Small utility helpers for normalizing dimensions and constructing shapes.
+"""
+
+from typing import Dict, List, Sequence, Union
+
+import torch
+
+from physicsnemo.utils.version_check import check_module_requirements
+
+check_module_requirements("physicsnemo.distributed.shard_tensor")
+
+from torch.distributed.tensor.placement_types import (  # noqa: E402
+    Shard,
+)
+
+from physicsnemo.distributed import ShardTensor  # noqa: E402
+
+aten = torch.ops.aten
+
+
+def unsqueeze_shape(shape: Union[torch.Size, Sequence[int]], dim: int) -> torch.Size:
+    """
+    Return a new `torch.Size` with a singleton dimension inserted at `dim`.
+
+    - If `dim` is within the current rank, the new dimension is inserted at that index.
+    - This mirrors the behavior of `torch.unsqueeze` at the shape level.
+
+    Args:
+        shape: The original shape as a `torch.Size` or sequence of integers.
+        dim: The dimension index at which to insert a singleton dimension.
+
+    Returns:
+        A new `torch.Size` with the inserted dimension.
+    """
+    o_shape = list(shape)
+    o_shape.insert(dim, 1)
+    return torch.Size(tuple(o_shape))
+
+
+def normalize_dim(dim: int, tensor_rank: int) -> int:
+    """
+    Normalize a possibly negative `dim` to a non-negative index for a given rank.
+
+    Follows PyTorch semantics for unsqueeze:
+    when `dim < 0`, the effective index is `tensor_rank + dim + 1`.
+
+    Args:
+        dim: The (possibly negative) dimension index.
+        tensor_rank: The rank (number of dimensions) of the tensor.
+
+    Returns:
+        The normalized non-negative dimension index.
+    """
+    while dim < 0:
+        dim = tensor_rank + dim + 1
+    return dim
```

Comment:
This function should just be:
```python
def normalize_dim(dim: int, tensor_rank: int) -> int:
    return dim if dim >= 0 else (dim % (tensor_rank + 1))
```
To avoid the `while` loop.